### ISSUE TITLE
Do not consider priorities when sorting workloads from different ClusterQueues

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes-sigs/kueue#1283]
**What would you like to be added**:

Use priorities to sort workloads within a ClusterQueue, but ignore the priorities when sorting the heads of multiple cluster queues within a cohort.

The design needs to avoid the race condition presented here https://github.com/kubernetes-sigs/kueue/pull/1024

**Why is this needed**:

In organizations where teams do not know each other, they might be incentivized to use higher priorities to always be ahead of the rest. Ignoring priorities across ClusterQueues would remove this incentive, while allowing users to use priorities within their ClusterQueue.

Maybe this calls for a setting in a new Cohort object.

**Completion requirements**:

This enhancement requires the following artifacts:

- [x] Design doc
- [x] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

**Comment by yaroslava-serdiuk** [comment]
Having that StrictFIFO is actually a priority queueing, I couldn't reproduce the bug that was described in https://github.com/kubernetes-sigs/kueue/pull/1024. Maybe there was no bug but actually not desired behavior when low priority workload was admitted and it was preempted just in the next scheduling cycle and we wanted to avoid such scenario and schedule higher priority workload in the first place? 

Anyway, my understanding is that we want to have the priority sorting in the scheduling loop as optional (enabled by default), so feature gate could be a solution here.

**Comment by alculquicondor** [comment]
The example above has some problems
but we can implement this much simpler case

![image](https://github.com/kubernetes-sigs/kueue/assets/1299064/3d8ebcd1-5a4a-4938-93c9-7d51e2d3db85)

**Comment by yaroslava-serdiuk** [comment]
I've added a test cases that you described https://github.com/kubernetes-sigs/kueue/issues/1283#issuecomment-1834370396 in https://github.com/kubernetes-sigs/kueue/pull/1399 and it turns out that the pending workload is still blocking borrowing. Note, that priorities doesn't matter here, I putted the highest priority for borrowing workload. The reason is that the pending workload has "Preempt" FlavorAssignmentMode and has no borrowing, so in sorting it goes before any workload that requires borrowing. 
It's definitely a bug, however it's not related to disabling priority feature. So I've added a feature gate to a separate PR https://github.com/kubernetes-sigs/kueue/pull/1406 and will look further how to fix this bug.

**Comment by alculquicondor** [comment]
For completeness when investigating a solution. In the following case, workload `#3` should be blocked, because `#2` is waiting to use its nominal quota.

![image](https://github.com/kubernetes-sigs/kueue/assets/1299064/7d3bcceb-eaf7-440e-ba56-0d39ce7acb47)

See this comment:

https://github.com/kubernetes-sigs/kueue/blob/e75090ab7ade6b55349768de8f3dff3c9becfaa4/pkg/scheduler/flavorassigner/flavorassigner.go#L201-L204

I think a potential solution could be to change this logic https://github.com/kubernetes-sigs/kueue/blob/e75090ab7ade6b55349768de8f3dff3c9becfaa4/pkg/scheduler/scheduler.go#L190 to only "use" (or block) up to the nominal quota.

**Comment by yaroslava-serdiuk** [comment]
Probably adding minimum from `workload resources` and `clusterQueue nominal resources that are left` [here](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/scheduler.go#L205) will help

**Comment by tenzen-y** [comment]
@yaroslava-serdiuk @alculquicondor I think this is almost done.

I guess that leaving task is adding docs for featureGate, `PrioritySortingWithinCohort`?
https://github.com/kubernetes-sigs/kueue/blob/main/site/content/en/docs/installation/_index.md#change-the-feature-gates-configuration

**Comment by alculquicondor** [comment]
yes, we should document this feature.

**Comment by kannon92** [comment]
@tenzen-y @yaroslava-serdiuk 

This feature is beta since 0.6. Any major issues with promoting this as GA?

**Comment by mimowo** [comment]
cc @mwielgus to keep me honest here, but IIRC this feature-gate is introduced as a tactical measure to allow users to set it to `false` for deployments where admins are concern about abusing priorities for scheduling (before introduction of fair sharing). Since the introduction of fair sharing the need to disable the feature should be gone, and we can make it GA. 

However, I would like to first notify some of our users and make sure they adopt fair sharing before this feature-gate is graduated (removed). So I propose to GA it in 0.12.

cc @gabesaba @mwysokin

**Comment by tenzen-y** [comment]
> cc [@mwielgus](https://github.com/mwielgus) to keep me honest here, but IIRC this feature-gate is introduced as a tactical measure to allow users to set it to `false` for deployments where admins are concern about abusing priorities for scheduling (before introduction of fair sharing). Since the introduction of fair sharing the need to disable the feature should be gone, and we can make it GA.
> 
> However, I would like to first notify some of our users and make sure they adopt fair sharing before this feature-gate is graduated (removed). So I propose to GA it in 0.12.
> 
> cc [@gabesaba](https://github.com/gabesaba) [@mwysokin](https://github.com/mwysokin)

+1
